File: src/apple/Packages/MapleUI/Sources/MapleUI/OrganismsB/MuiTemplateGalleryModal.swift
// MuiTemplateGalleryModal.swift — Maple UI Organisms · Modals (unified-
// component-catalog.md §4.4). Browse and apply an edit template, built on
// Overlay Shell from Search Bar, Card, Empty State.

import SwiftUI

public struct MuiGalleryTemplate: Identifiable, Sendable {
    public let id: String
    public let name: String
    public let thumbnailUrl: URL?
    public let description: String?
    public let category: String?

    public init(id: String, name: String, thumbnailUrl: URL?, description: String? = nil, category: String? = nil) {
        self.id = id
        self.name = name
        self.thumbnailUrl = thumbnailUrl
        self.description = description
        self.category = category
    }
}

public struct MuiTemplateGalleryModal: View {
    public let isPresented: Bool
    public let contained: Bool
    public let templates: [MuiGalleryTemplate]
    @Binding public var search: String
    public let templateApplied: ((MuiGalleryTemplate) -> Void)?
    public let dismissed: (() -> Void)?

    public init(
        isPresented: Bool,
        contained: Bool = false,
        templates: [MuiGalleryTemplate],
        search: Binding<String>,
        templateApplied: ((MuiGalleryTemplate) -> Void)? = nil,
        dismissed: (() -> Void)? = nil
    ) {
        self.isPresented = isPresented
        self.contained = contained
        self.templates = templates
        self._search = search
        self.templateApplied = templateApplied
        self.dismissed = dismissed
    }

    private var filtered: [MuiGalleryTemplate] {
        Self.filtered(templates, search: search)
    }

    public var body: some View {
        MuiOverlayShell(isPresented: isPresented, size: .lg, accessibilityLabel: "Template Gallery", contained: contained) {
            MuiText("Template Gallery", variant: .sheetTitle)
        } content: {
            VStack(alignment: .leading, spacing: MuiTokens.spacingMd) {
                MuiSearchBar(value: $search, placeholder: "Search templates…")

                if filtered.isEmpty {
                    MuiEmptyState(icon: "square.grid.2x2", title: "No templates found")
                } else {
                    let columns = [GridItem(.adaptive(minimum: 140), spacing: MuiTokens.spacingSm)]
                    LazyVGrid(columns: columns, spacing: MuiTokens.spacingSm) {
                        ForEach(filtered) { template in
                            MuiCard(
                                url: template.thumbnailUrl, alt: template.name, title: template.name,
                                subtitle: template.category, pressed: { templateApplied?(template) }
                            )
                        }
                    }
                }
            }
        } footer: {
            HStack {
                Spacer()
                MuiButton(label: "Close", variant: .ghost) { dismissed?() }
            }
        } dismissed: {
            dismissed?()
        }
    }

    /// Templates whose name, description, or category matches `search`,
    /// case-insensitively — all templates when `search` is blank. Public +
    /// static so this is unit-testable without rendering a view.
    public static func filtered(_ templates: [MuiGalleryTemplate], search: String) -> [MuiGalleryTemplate] {
        let query = search.trimmingCharacters(in: .whitespaces).lowercased()
        guard !query.isEmpty else { return templates }
        return templates.filter {
            let haystack = "\($0.name) \($0.description ?? "") \($0.category ?? "")".lowercased()
            return haystack.contains(query)
        }
    }
}

#Preview("MuiTemplateGalleryModal") {
    struct Demo: View {
        @State private var open = false
        @State private var search = ""
        var body: some View {
            ZStack {
                MuiTokens.bg
                MuiButton(label: "Open Template Gallery", variant: .primary) { open = true }
                MuiTemplateGalleryModal(
                    isPresented: open,
                    templates: [
                        MuiGalleryTemplate(id: "1", name: "Moody Landscape", thumbnailUrl: nil, category: "Landscape"),
                        MuiGalleryTemplate(id: "2", name: "Portrait Warm", thumbnailUrl: nil, category: "Portrait"),
                    ],
                    search: $search,
                    dismissed: { open = false }
                )
            }
            .frame(width: 420, height: 380)
        }
    }
    return Demo()
}

File: src/windows/Maple.WinUI/MapleUI/Organisms/MuiTemplateGalleryModal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>One browsable template.</summary>
    public sealed record MuiGalleryTemplate(string Id, string Title, string? Subtitle = null, ImageSource? Preview = null);

    /// <summary>
    /// Maple.UI Template Gallery modal organism (unified-component-catalog.md
    /// §4.4, "Template Gallery" row: "Browse and apply templates", built
    /// from Card, Search Bar, Empty State) — a <see cref="MuiSearchBar"/>
    /// filters a wrapped grid of <see cref="MuiCard"/>s, or a
    /// <see cref="MuiEmptyState"/> when nothing matches.
    /// </summary>
    public sealed class MuiTemplateGalleryModal : ContentControl
    {
        public static readonly DependencyProperty IsOpenProperty =
            DependencyProperty.Register(nameof(IsOpen), typeof(bool), typeof(MuiTemplateGalleryModal),
                new PropertyMetadata(false, (d, e) => ((MuiTemplateGalleryModal)d)._shell.IsOpen = (bool)e.NewValue));

        public static readonly DependencyProperty ContainedProperty =
            DependencyProperty.Register(nameof(Contained), typeof(bool), typeof(MuiTemplateGalleryModal),
                new PropertyMetadata(false, (d, e) => ((MuiTemplateGalleryModal)d)._shell.Contained = (bool)e.NewValue));

        public static readonly DependencyProperty TemplatesProperty =
            DependencyProperty.Register(nameof(Templates), typeof(IReadOnlyList<MuiGalleryTemplate>), typeof(MuiTemplateGalleryModal),
                new PropertyMetadata(null, (d, _) => ((MuiTemplateGalleryModal)d).Rebuild()));

        public bool IsOpen { get => (bool)GetValue(IsOpenProperty); set => SetValue(IsOpenProperty, value); }
        public bool Contained { get => (bool)GetValue(ContainedProperty); set => SetValue(ContainedProperty, value); }

        public IReadOnlyList<MuiGalleryTemplate>? Templates
        {
            get => (IReadOnlyList<MuiGalleryTemplate>?)GetValue(TemplatesProperty);
            set => SetValue(TemplatesProperty, value);
        }

        public event EventHandler? Dismissed;
        public event EventHandler<string>? TemplateApplied;

        private readonly MuiOverlayShell _shell = new() { Size = MuiOverlayShellSize.Lg, AriaLabel = "Template Gallery" };
        private readonly MuiSearchBar _search = new() { Placeholder = "Search templates…" };
        private readonly StackPanel _rows = new() { Orientation = Orientation.Vertical, Spacing = 10 };
        private readonly MuiEmptyState _empty = new() { IconName = "grid-lg", Title = "No templates match" };
        private readonly MuiButton _close = new() { Variant = MuiButtonVariant.Ghost, Label = "Close" };
        private string _filter = string.Empty;

        public MuiTemplateGalleryModal()
        {
            var body = new StackPanel { Orientation = Orientation.Vertical, Spacing = 14 };
            body.Children.Add(_search);
            body.Children.Add(_rows);
            body.Children.Add(_empty);

            var footer = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8, HorizontalAlignment = HorizontalAlignment.Right };
            footer.Children.Add(_close);

            _shell.Header = new MuiText { Text = "Template Gallery", Variant = MuiTextVariant.SheetTitle };
            _shell.Body = body;
            _shell.Footer = footer;
            Content = _shell;
            HorizontalContentAlignment = HorizontalAlignment.Stretch;
            VerticalContentAlignment = VerticalAlignment.Stretch;

            _shell.Dismissed += (_, _) => { IsOpen = false; Dismissed?.Invoke(this, EventArgs.Empty); };
            _close.Click += (_, _) => { IsOpen = false; Dismissed?.Invoke(this, EventArgs.Empty); };
            _search.Committed += (_, text) => { _filter = text; Rebuild(); };

            Rebuild();
        }

        private void Rebuild()
        {
            var matches = (Templates ?? Array.Empty<MuiGalleryTemplate>())
                .Where(t => string.IsNullOrEmpty(_filter) || t.Title.Contains(_filter, StringComparison.OrdinalIgnoreCase))
                .ToList();

            _empty.Visibility = matches.Count == 0 ? Visibility.Visible : Visibility.Collapsed;
            _rows.Visibility = matches.Count > 0 ? Visibility.Visible : Visibility.Collapsed;

            _rows.Children.Clear();
            const int perRow = 3;
            StackPanel? row = null;
            for (var i = 0; i < matches.Count; i++)
            {
                if (i % perRow == 0)
                {
                    row = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 10 };
                    _rows.Children.Add(row);
                }
                var template = matches[i];
                var card = new MuiCard { Title = template.Title, Subtitle = template.Subtitle, Source = template.Preview, Width = 140 };
                card.Pressed += (_, _) => TemplateApplied?.Invoke(this, template.Id);
                row!.Children.Add(card);
            }
        }
    }
}
